api/src/test/java/org/pmiops/workbench/ras/RasLinkServiceTest.java

```diff
@@ -0,0 +1,167 @@
+
+@RunWith(SpringRunner.class)
+@DataJpaTest
+public class RasLinkServiceTest {
+  private static final Timestamp NOW = Timestamp.from(Instant.now());
+  private static final FakeClock CLOCK = new FakeClock(NOW.toInstant(), ZoneId.systemDefault());
+
+  private static final String AUTH_CODE = "code";
+  private static final String REDIRECT_URL = "url";
+  private static final String ACCESS_TOKEN = "access_token_1";
+  private static final String LOGIN_GOV_USERNAME = "foo@gmail.com";
+  private static final String USER_INFO_JSON_LOGIN_GOV =
+      "{\"preferred_username\":\"user1@Login.Gov.com\", \"email\":\"" + LOGIN_GOV_USERNAME + "\"}";
+  private static final String USER_INFO_JSON_ERA =
+      "{\"preferred_username\":\"user2@eraCommons.com\", \"email\":\"" + LOGIN_GOV_USERNAME + "\"}";
+  private static final String ID_TOKEN_JWT_IAL_1 =
+      JWT.create()
+          .withClaim(
+              ACR_CLAIM,
+              "https://stsstg.nih.gov/assurance/ial/1 https://stsstg.nih.gov/assurance/aal/1")
+          .sign(Algorithm.none());
+  private static final String ID_TOKEN_JWT_IAL_2 =
+      JWT.create()
+          .withClaim(
+              ACR_CLAIM,
+              "https://stsstg.nih.gov/assurance/ial/2 https://stsstg.nih.gov/assurance/aal/1")
+          .sign(Algorithm.none());
+  private static final TokenResponse TOKEN_RESPONSE_IAL1 =
+      new TokenResponse().setAccessToken(ACCESS_TOKEN).set(Id_TOKEN_FIELD_NAME, ID_TOKEN_JWT_IAL_1);
+  private static final TokenResponse TOKEN_RESPONSE_IAL2 =
+      new TokenResponse().setAccessToken(ACCESS_TOKEN).set(Id_TOKEN_FIELD_NAME, ID_TOKEN_JWT_IAL_2);
+
+  private long userId;
+  private static DbUser currentUser;
+
+  private RasLinkService rasLinkService;
+  private ObjectMapper objectMapper = new ObjectMapper();
+
+  @Autowired private UserService userService;
+  @Autowired private UserDao userDao;
+  @Mock private OpenIdConnectClient mockOidcClient;
+
+  @TestConfiguration
+  @Import({
+    UserServiceTestConfiguration.class,
+  })
+  static class Configuration {
+    @Bean
+    Clock clock() {
+      return CLOCK;
+    }
+
+    @Bean
+    Random random() {
+      return new FakeLongRandom(123);
+    }
+
+    @Bean
+    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    public WorkbenchConfig getWorkbenchConfig() {
+      return WorkbenchConfig.createEmptyConfig();
+    }
+
+    @Bean
+    @Scope("prototype")
+    DbUser user() {
+      return currentUser;
+    }
+  }
+
+  @MockBean UserServiceAuditor mockUserServiceAuditor;
+  @MockBean FireCloudService mockFireCloudService;
+  @MockBean ComplianceService mockComplianceService;
+  @MockBean DirectoryService mockDirectoryService;
+  @MockBean private FreeTierBillingService mockFreeTierBillingService;
+  @Mock private OpenIdConnectClient mockRasOidcClient;
+
+  @Before
+  public void setUp() throws Exception {
+    rasLinkService = new RasLinkService(userService, mockRasOidcClient);
```

Comment:
In an ideal test case, I think you would set up all the dependencies, and then `@Autowire` in your target service. To have that work in this case you would need to put `mockRasOidcClient` into the test configuration.

Reply:
Done.
But another pattern that we use before:
is if the target service does not need bean, I just need @Autowire the target service.
But if we need to mock some service, or we want to change the target service's parameter. Then `@Autowire` or `@MockBean` all target's service parameters, then construct the target service on our own. This makes some tests easier and more flexible in some cases.

And also, I am not sure @TestConfiguration is a good candidate in the future, that also confused me.